src/object/srv_obj.c

```diff
@@ -41,6 +41,273 @@
+
+	if (!dta->dta_share) {
+		if (rc > 0) {
+			/* Only trigger aggregation in separated ULT. */
+			while (1) {
+				rc = vos_dtx_aggregate(cont->sc_hdl);
+				if (rc != DTX_ACT_AGGREGATE)
+					break;
+
+				ABT_thread_yield();
+			}
+		}
+
+		if (!cont_hdl->sch_cont)
+			ds_cont_put(cont);
+		ds_cont_hdl_put(cont_hdl);
+	}
+	D_FREE(dta);
+}
+
+static void
+ds_tx_aggregate(void *arg)
+{
+	struct ds_tx_args	*dta = (struct ds_tx_args *)arg;
+
+	D_ASSERT(!dta->dta_share);
+
+	while (1) {
+		int	rc;
+
+		rc = vos_dtx_aggregate(dta->dta_cont->sc_hdl);
+		if (rc != DTX_ACT_AGGREGATE)
+			break;
+
+		ABT_thread_yield();
+	}
+
+	if (!dta->dta_cont_hdl->sch_cont)
+		ds_cont_put(dta->dta_cont);
+	ds_cont_hdl_put(dta->dta_cont_hdl);
+	D_FREE(dta);
+}
+
+static void
+ds_tx_abort(uuid_t po_uuid, uuid_t co_uuid, struct daos_tx_id *dti,
+	    daos_unit_oid_t *oid, uint32_t version)
+{
+	struct daos_tx_entry	 dte;
+
+	dte.dte_xid = *dti;
+	dte.dte_oid = *oid;
+	dtx_abort(po_uuid, co_uuid, &dte, 1, version);
+}
+
+static int
+ds_tx_post_handle(struct ds_cont_hdl *cont_hdl, struct ds_cont *cont,
+		  struct daos_tx_handle *dth, struct daos_tx_id *dti,
+		  daos_unit_oid_t *oid, struct daos_tx_id *dti_cos,
+		  int dti_cos_count, int rc, bool rpc_dispatched,
+		  bool inline_dti_cos, bool leader)
+{
+	int	err;
+
+	if (dti_cos != NULL) {
+		if (rpc_dispatched && rc < 0) {
+			/* XXX: Even if we fail to make related modification
+			 *	for some reason, we still need to commit the
+			 *	DTXs for CoS. Because other replica may have
+			 *	already committed them.
+			 */
+			err = vos_dtx_commit(cont->sc_hdl, dti_cos,
+					     dti_cos_count);
+			D_ERROR(DF_UOID" can not commit DTX "DF_DTI
+				" for CoS: rc = %d\n",
+				DP_UOID(*oid), DP_DTI(dti), err);
+		}
+
+		if (!inline_dti_cos)
+			D_FREE(dti_cos);
+	}
+
+	/* If the DTX is started befoe DTX resync operation (for rebuild),
+	 * then it is possbile that the DTX resync ULT may have aborted
+	 * current DTX before remote replica(s) modification by race. So
+	 * let's check DTX status locally before marking as 'committable'.
+	 */
+	if (leader && rc >= 0 && dti->dti_sec <= cont->sc_dtx_resync_time) {
+		err = vos_dtx_check_committable(cont->sc_hdl, NULL, dti, 0,
+						false);
+		if (err < 0)
+			rc = (err == -DER_NONEXIST ? -DER_INPROGRESS : err);
+	}
+
+	err = vos_dtx_end(dth, rc, leader);
+	switch (err) {
+	case DTX_ACT_AGGREGATE: {
+		struct ds_tx_args *dta;
+
+		dta = ds_tx_prepare_args(cont_hdl, cont, false);
+		if (dta == NULL)
+			D_WARN("Not enough RAM for DTX aggregation.\n");
+		else
+			dss_ult_create(ds_tx_aggregate, dta, DSS_ULT_SELF,
+				dss_get_module_info()->dmi_tgt_id, 0, NULL);
+		break;
+	}
+	case DTX_ACT_COMMIT_ASYNC:
+		if (!cont->sc_tx_committing) {
+			struct ds_tx_args *dta;
+
+			dta = ds_tx_prepare_args(cont_hdl, cont, false);
+			if (dta == NULL)
+				D_WARN("Not enough RAM for DTX commit.\n");
+			else
+				dss_ult_create(ds_tx_commit, dta, DSS_ULT_SELF,
```

Comment:
Do we need multiple commit ULTs for same container?

Reply:
Not need, that is why I introduced the sc_tx_committing flag.